### PR TITLE
dry: merge dm's {alertmanager,grafana,prometheus} into cluster

### PR DESCRIPTION
**Pull request by 9547**
### What problem does this PR solve? 

Try to implement #853 

### What is changed and how it works?

The logic of dm's monitor components almost the same as cluster's, merge them into one place, less code, easier to maintain.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
merge dm's {alertmanager,grafana,prometheus} into cluster
```

### CODE REVIEW

**Comment by 9547** [comment]
@lucklove PTAL, thanks

File: pkg/cluster/spec/spec.go [hunk 1]
```diff
@@ -192,6 +194,11 @@ func (s *Specification) GetMonitoredOptions() *MonitoredOptions {
 	return &s.MonitoredOptions
 }
 
+// GetGlobalOptions implements Topology interface.
+func (s *Specification) GetGlobalOptions() GlobalOptions {
```
Comment (lucklove) [inline pkg/cluster/spec/spec.go:198]: This function is dup with `BaseTopo().GlobalOptions`

File: pkg/cluster/spec/alertmanager.go [hunk 2]
```diff
@@ -62,7 +63,7 @@ func (s AlertManagerSpec) IsImported() bool {
 }
 
 // AlertManagerComponent represents Alertmanager component.
-type AlertManagerComponent struct{ *Specification }
+type AlertManagerComponent struct{ Topology }
```
[the remark this answers is not in the document]
Comment (9547) [reply pkg/cluster/spec/alertmanager.go:66]: IMO, `Topology` is an interface which could hold for implementation is the concrete `Component`, so I change `AlertManagerComponent`, `MonitorComponent` and `GrafanaComponent` into this style, there are little differene between cluster and dm's topology.  While `Specification` stands for the specific `Component`, such as TiDB, TiKV, PD... So the other Components changed to `*Specification`

File: pkg/cluster/spec/alertmanager.go [hunk 3]
```diff
@@ -76,8 +77,15 @@ func (c *AlertManagerComponent) Role() string {
 
 // Instances implements Component interface.
 func (c *AlertManagerComponent) Instances() []Instance {
-	ins := make([]Instance, 0, len(c.Alertmanager))
-	for _, s := range c.Alertmanager {
+	val, found := findSliceField(c.Topology, "Alertmanager")
```
[the remark this answers is not in the document]
Comment (9547) [reply pkg/cluster/spec/alertmanager.go:80]: Good idea 👍

File: pkg/cluster/spec/grafana.go [hunk 4]
```diff
@@ -160,13 +164,22 @@ func (i *GrafanaInstance) InitConfig(
 		return err
 	}
 
+	topo := reflect.ValueOf(i.topo)
+	if topo.Kind() == reflect.Ptr {
+		topo = topo.Elem()
+	}
+	val := topo.FieldByName("Monitors")
+	if (val == reflect.Value{}) {
```
Comment (AstroProfundis) [inline pkg/cluster/spec/grafana.go:172]: Is this check necessary? There is check for monitor list length just below.

Comment (9547) [reply pkg/cluster/spec/grafana.go:172]: This sentinel checks if the struct contains field `Monitors` or not, if not, then `val.Interface()` will panic

**Review by AstroProfundis — APPROVED** [review]

**Pull request merged**